src/backend/metal/src/command.rs

```diff
@@ -1037,13 +1039,84 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
         // TODO: MTLRenderCommandEncoder.textureBarrier on macOS?
     }
 
-    fn fill_buffer(
+    fn fill_buffer<R>(
         &mut self,
-        _buffer: &native::Buffer,
-        _range: Range<buffer::Offset>,
-        _data: u32,
-    ) {
-        unimplemented!()
+        buffer: &native::Buffer,
+        range: R,
+        data: u32,
+    ) where
+        R: RangeArg<u64>,
+    {
+        let inner = unsafe { &mut *self.inner.get() };
+        let sink = inner.sink.as_mut().unwrap();
+
+        let mut pipes = self.shared.service_pipes
+            .lock()
+            .unwrap();
+
+        let pso = pipes
+            .get_fill_buffer(&self.shared.device)
+            .to_owned();
+
+        let array = [data];
+
+        // TODO: Reuse buffer allocation
+        let fill = self.shared.device.new_buffer_with_data(
+            array.as_ptr() as *const _,
+            mem::size_of::<u32>() as _,
+            metal::MTLResourceOptions::StorageModeShared,
+        );
+
+        inner.retained_buffers.push(fill.clone());
+
+        let start = *range.start().unwrap();
+        assert_eq!(start & 3, 0);
+
+        let end = match range.end() {
+            Some(e) => {
+                assert_eq!(e & 3, 0);
+                *e
+            },
+            None =>  (buffer.raw.length() - buffer.offset) & !3,
```

Comment:
I thought the point of the quote was to clamp the value down to the nearest multiple of 4 to ensure a complete write? What happens if we try to fill another 32-bits and only have 16-bits remaining?

Reply:
The quote is from the description of `size`, I don't read it as talking about the value to clamp. It seems pretty clear from that section that filling the whole buffer should be supported even if the length is not a multiple of 4, but I still don't understand the "nearest smaller multiple is used" part...

btw, here is some contradicting info from the [Vulkan Programming Guide](https://books.google.ca/books?id=XVBxDQAAQBAJ&pg=PT134&lpg=PT134&dq=vkCmdFillBuffer+non+multiple+of+4&source=bl&ots=zDRBS9s_Ed&sig=gb8teKy6_UJG61YC_ztJahXcRTA&hl=en&sa=X&ved=0ahUKEwiMzYnG4_PaAhUQuVkKHdUCC6EQ6AEILTAB#v=onepage&q=vkCmdFillBuffer%20non%20multiple%20of%204&f=false)

Reply:
Right, so in the case when `VK_WHOLE_SIZE` is provided, we receive `None` here. If the remaining size (`buffer.raw.length() - buffer.offset`) not word-aligned, we round down to the nearest smaller multiple of 4 (i.e. rounding down to the nearest word boundary) and use that as the size. But we don't assert here, so it would be fine to have a size that isn't word-aligned (only in the `None` case). Is this incorrect?

I agree that the wording is a bit inconsistent..

Reply:
Oooh, I see. Now it makes sense. So the `cmdFillBuffer` will never touch the remainder 1-3 bytes of non-aligned size. That makes our life easier, and also relaxes the necessity of CTS changes.